### PR TITLE
feat(ai): add structured decision client

### LINKED ISSUE #214: C1 — Build a provider-agnostic Jev structured-decision client
# C1 — Build a provider-agnostic Jev structured-decision client

## Context

Child C1 of [Epic C — Multi-candidate classifiers](https://github.com/vouchington/vouchington/issues/177). This is concrete implementation work for the AI Platform initiative.

## Blocked by

- None.

Outcome: a typed client supports Noul, Choice, and Score requests while preserving
provider probabilities, confidence, raw responses, and per-question identity end
to end.

Owned paths:

- New backend/modules/structured-decisions/**
- backend/modules/utils/http-dispatchers.mts
- Existing provider boundaries in backend/modules/openai-utils/**
- Deterministic tests beside the new module and a dedicated credentialed test suite
  for live-provider smoke contracts

Implementation scope:

- Implement direct TypeSafe System One transport.
- Implement the OpenRouter Jev transport only when its response preserves the
  required probability and confidence data.
- Expose a provider-agnostic interface for Noul, Choice, and Score.
- Validate every response strictly: question identity, result cardinality,
  probability domains, Choice normalization, confidence domains, and raw-response
  retention.
- Preserve Noul probability; do not fabricate confidence when the provider does
  not emit it.
- Retry only the same Jev model through the configured transport/provider retry
  policy.
- Verify maximum usable question capacity empirically using the active
  transport/model limits rather than hard-coding a stale total-context
  assumption.
- Benchmark the fixed moderation set and realistic large tagging/community
  candidate sets, including cache-hit assumptions.

Guardrails and out of scope:

- No automatic fallback to a different model.
- Missing, malformed, partial, or ambiguous results must fail closed.
- No classifier writes, votes, clearance, tagging, unpublishing, or story
  creation belong here.
- Do not hand-edit generated contracts or schema snapshots.
- Credentials remain external to source, fixtures, logs, and commits.

Acceptance criteria and tests:

- Deterministic unit and transport tests cover valid Noul, Choice, and Score
  responses; malformed JSON; missing, duplicate, and unknown answers;
  out-of-range probabilities; provider errors; timeout; retry; status handling;
  and raw-response preservation.
- Add a dedicated credentialed suite containing exactly one or two live
  OpenRouter/Jev smoke-contract tests. They verify only the active integration
  contract needed to preserve provider probability and confidence data; they do
  not replace deterministic parsing, validation, failure, or retry coverage.
- An empirical benchmark records the question ceiling and probability stability
  as candidate count grows.
- The cost benchmark includes Jev versus the current moderation/tagging traffic
  shapes.
- The focused backend Vitest suite and the credentialed smoke suite pass when
  their external credential is configured.

### LINKED ISSUE #242: Plan: Build a provider-agnostic Jev structured-decision client
## Solves

- #214
- Source: https://github.com/vouchington/vouchington/issues/214

## Why

- Goal: provide one strict, provider-neutral structured-decision boundary for Jev Noul, Choice, and Score calls so later classifier work can batch candidates without depending on provider envelopes.
- Root cause: the repository has no System One client, while the two viable transports expose similar but independently versioned HTTP contracts. Treating either response as ordinary generated text would discard the calibrated probabilities and confidence that make the classifier design safe.
- Chosen fix: add a data-store-free `@modules/structured-decisions` package with explicit TypeSafe and OpenRouter Decisions transports, transport-specific strict decoding into one discriminated result model, raw-payload retention, and bounded same-model retries. Select the transport explicitly; never switch transport or model after an ambiguous/billable attempt. OpenRouter's dedicated `/api/alpha/decisions` endpoint is used because a credentialed probe returned native Noul, Choice, and Score payloads, so chat-completions emulation and an additional SDK are unnecessary.

## KPIs

| KPI | Target | Measurement |
| --- | ------ | ----------- |
| Primitive coverage | Noul, Choice, and Score preserve provider-native decision fields and question identity | Deterministic contract tests plus one live OpenRouter/Jev smoke test containing all three primitives |
| Fail-closed validation | 100% of missing, duplicate, unknown, malformed, out-of-range, or non-normalized answers reject before returning a result | Mutation-sensitive deterministic validation tests |
| Retry safety | Only retryable network/timeout/429/529/5xx failures repeat the same model on the explicitly selected transport; malformed results and non-retryable 4xx responses make one attempt | Deterministic attempt-count and abort tests |
| Credential isolation | Exactly one credentialed OpenRouter test, in its own Vitest project; no key appears in source, fixtures, output, or committed artifacts | Project ownership/workflow tests, secret inventory, and diff scan |
| Capacity evidence | Repeated fixed-anchor trials across representative 6-question moderation and increasing dynamic sets distinguish a measured provider failure from merely succeeding through an operator cap | Opt-in benchmark run against `typesafe/jev-1.13`; exact observations stay in a mode-0600 private artifact and only non-sensitive method/conclusions enter public docs |
| Cost evidence | Jev cost is calculated from observed input usage for fixed and dynamic sets and compared with the documented current moderation/tagging shapes | Deterministic cost-model tests; real usage and per-unit cost stay in the private documentation tier |

## Alternatives analysis

| Alternative | Benefits | Costs or risks | Decision reason |
| ----------------------- | -------- | -------------- | ---------------------------- |
| No change: continue per-agent scoring calls | No new provider boundary | Preserves candidate fan-out, repeated inputs, tool loops, and uncalibrated generated-text parsing | Rejected because it does not solve #214 or enable C3-C10 |
| Reuse: send Jev through the existing OpenAI Responses utilities or OpenRouter chat completions | Reuses an established SDK and retry code | The wire contract is not OpenAI Responses; text/tool-call emulation can lose native probabilities and confidence | Rejected because the live OpenRouter Decisions endpoint already exposes the native contract and strict decoding is safer |
| Reuse: add `@openrouter/sdk` and build only an OpenRouter client | Generated alpha-endpoint types | Adds a beta generated dependency, omits the required direct TypeSafe transport, and still requires runtime validation for remote data | Rejected because the repository's existing `undici` dispatcher plus local validation supplies the required boundary without another SDK |
| Materially different: implement two explicit raw-HTTP adapters behind one typed client | Small dependency surface, preserves each provider envelope, supports direct TypeSafe and the proven OpenRouter native API | Requires maintaining two decoders and local retry logic | Accepted because provider differences remain visible at the boundary while callers receive one strict model |

## Affected files and modules

| Path or module | Existing or new | Role and change | Dependencies | Dependents |
| -------------- | --------------- | --------------- | ------------ | ---------- |
| `backend/modules/structured-decisions/**` | New | Public types/client, TypeSafe and OpenRouter transports, strict decoding, same-model retry policy, tests, benchmark CLI, package metadata, and package README | `@modules/utils`, `undici` | C3 multi-candidate call layer and later classifiers |
| `backend/modules/utils/http-dispatchers.mts` | Existing | Reuse the long-running external fetch dispatcher; change only if a narrow injectable fetch seam is required by the final transport implementation | `undici` | Existing external integrations and new structured-decision transports |
| `backend/catalogs/README.md` and `pnpm-lock.yaml` | Existing | Register the new workspace package and generated importer state | New package manifest | Monorepo package discovery and CI |
| `vitest.config.mts`, `ci/vitest/project-ownership.mts`, `ci/run-vitest-project-group.mts`, `.no-mistakes.yml`, `.github/workflows/VITEST.md` | Existing | Add a dedicated `backend-openrouter` credentialed project, exclude its filename pattern from `backend-modules`, add ownership/routing and integration mapping, and regenerate project documentation | New `*.openrouter.test.mts` contract test | CI selection, coverage, and local project runners |
| `.github/workflows/tests-backend-credentialed.yml`, `.github/workflows/ci.yml`, `.github/ci-path-filters.yml`, `.github/workflows/workflow-secrets-inventory.mts` | Existing | Forward, preflight, and expose only `OPENROUTER_API_KEY` to the credentialed backend job; select the new project when its owned paths change | Repository secret and Vitest ownership | Trusted CI runs and secret-policy tests |
| `ci/transient-retry/backend-credentialed-*`, `ci/transient-retry/reference-cautionary-examples.md`, `docs/development/reference-ci-classifying-transient-infrastructure-failures.md`, `docs/development/reference-ci-test-workflows.md` | Existing | Extend the credentialed-project topology and narrowly classify OpenRouter transport timeouts/429/5xx while continuing to reject assertion, schema, auth, and mixed failures | New project include glob and live-test error boundary | Main CI retry classification and operator docs |
| `.env.example`, `docs/development/local-env-vars.md`, `docs/overview/infrastructure/reference-environment-variables-ai-ml.md` | Existing | Document local and CI credential setup without values; direct TypeSafe remains caller-injected because no TypeSafe repository credential exists | External credentials | Developers and CI operators |
| `docs/overview/architecture/structured-decisions.md` | New | Durable provider contract, invariants, benchmark method, non-sensitive conclusions, cost-model assumptions, and limitations; exact operational observations remain private | Implemented client and sanitized benchmark conclusion | C2-C14 architecture work |

## Before and after

```mermaid
flowchart LR
  CallerBefore[Future classifier caller] --> ProviderEnvelope[Provider-specific or generated-text response]
  ProviderEnvelope --> Risk[Probability/confidence loss or permissive parsing]
  CallerAfter[Future classifier caller] --> Client[Structured-decision client]
  Client --> TypeSafe[TypeSafe System One]
  Client --> OpenRouter[OpenRouter Decisions]
  TypeSafe --> Validator[Strict transport decoder]
  OpenRouter --> Validator
  Validator --> Typed[Exact typed answers plus raw response]
```

| Concern | Before | After |
| ------- | ------ | ----- |
| Transport | No repository client for System One | Explicit direct TypeSafe or OpenRouter Decisions transport |
| Provider selection | Undefined | Caller selects one transport; no automatic cross-transport or cross-model fallback |
| Response trust | No runtime contract | Exact identity/cardinality/type/domain/normalization validation, failing closed |
| Retries | Undefined | Bounded same-model retries only for unambiguous retryable failures and abort-aware delays |
| Live compatibility | Ad hoc local probe | One isolated credentialed OpenRouter/Jev smoke contract in `backend-openrouter` |
| Capacity and cost | Launch-price assumptions plus inconsistent documented limits (TypeSafe: 64k total request and 32k state-plus-longest-question; OpenRouter catalog: 32k context) | Reproducible per-transport representative benchmark with dated observed usage and no hard-coded question-count ceiling |

## Implementation plan

1. Add failing deterministic tests first for request identity, all three valid primitive shapes, raw response retention, exact answer coverage, duplicate/unknown/missing answers, invalid domains/normalization, malformed JSON, aborts, status classification, and bounded attempt counts.
2. Define a discriminated transport-independent request/result model whose types make primitive-specific fields explicit: Noul has probability but no fabricated confidence; Choice has selected option, exact option probabilities, and confidence; Score has score, legend/probabilities, and confidence.
3. Validate outgoing state/questions before dispatch, then implement separate strict decoders for TypeSafe System One and OpenRouter Decisions. Validate question IDs and primitive type against the request, require one answer per requested ID, require Choice selections to belong to the requested criteria, require Score values within `[0, criteria.length - 1]`, require exact Score legend/probability keys and descriptions, validate finite `[0,1]` probabilities/confidence and Choice/Score probability sums within a documented numeric tolerance, and retain the unmodified parsed provider body with normalized results.
4. Implement direct `undici` transports using the long-running external dispatcher: `POST https://api.typesafe.ai/v1/systemone` with caller-injected TypeSafe credentials and `jev-latest`, and `POST https://openrouter.ai/api/alpha/decisions` with caller-injected OpenRouter credentials, `typesafe/jev-1.13`, and an explicit TypeSafe-only/no-fallback provider preference. Mark only the direct provider call functions as integration seams.
5. Add an abort-aware, bounded retry wrapper that preserves the exact transport, endpoint, request, and Jev model. Honor valid `Retry-After` guidance within a cap; retry network/timeout/429/529/5xx failures; never retry malformed success responses, ordinary 4xx responses, or after caller cancellation.
6. Add the workspace package metadata, public barrel, catalog entry, and generated lockfile update. Keep classifier persistence, sharding, thresholds, votes, and side effects out of the module.
7. Add exactly one `*.openrouter.test.mts` live smoke test in a new `backend-openrouter` Vitest project. It submits one request containing Noul, Choice, and Score, verifies native probability/confidence preservation and exact IDs, and never prints request headers or credential values.
8. Wire `OPENROUTER_API_KEY` through the trusted credentialed backend workflow, caller, path filters, project ownership, secret inventory, and test-selection mappings; regenerate `VITEST.md`. Keep the new project separate from deterministic `backend-modules` tests, and make missing credentials fail its explicit local/CI invocation rather than skip.
9. Extend credentialed transient classification to recognize only OpenRouter timeouts/429/5xx failures owned by the new project, with synthetic-log tests and updated operator documentation. Authentication, parser/schema assertions, and mixed failures remain non-retryable.
10. Add an opt-in benchmark CLI that builds representative fixed and dynamic candidate sets, runs repeated fixed-anchor trials as size increases under an explicit operator cap, and reports whether it observed a first provider failure or only succeeded through the cap. Add deterministic cost calculations for Jev and current comparison price/cache-hit assumptions.
11. Source `~/voucha.env`, run the one live test explicitly, and prove Vitest executed one test rather than skipping it. Run the benchmark once with the configured OpenRouter credential, retain exact capacity/drift/latency/usage/cost observations only in a mode-0600 private artifact for the private documentation tier, and publish only architecture, method, and non-sensitive conclusions in this repository.

## Affected tests

| Test | Existing or new | Why affected | Behavior to check before implementation |
| ---- | --------------- | ------------ | --------------------------------------- |
| `ci/vitest/project-ownership.test.mts` and `.github/workflows/vitest-project-ownership.test.mts` | Existing | The credentialed workflow gains a project and credential owner | Existing four projects and generated ownership table agree before adding the fifth |
| `ci/run-vitest-project-group.test.mts` | Existing | Local named project routing gains `backend-openrouter` | Current credentialed projects are accepted and unknown projects reject |
| `ci/no-mistakes-config-freshness.test.mts` and `ci/no-mistakes-test-plan-config.test.mts` | Existing | Integration-suite and selection config gain an OpenRouter mapping | Current config is aligned before extending it |
| `.github/workflows/workflow-secrets-policy.test.mts` and secret-context tests | Existing | A new repository secret crosses the trusted reusable-workflow boundary | Existing secrets are inventoried, guarded, and forwarded only under trusted context |
| `.github/workflows/__tests__/vitest-ci-triggers*.test.mts` | Existing | New module and live-test paths must select the right deterministic and credentialed projects | Current path filters select existing credentialed projects correctly |
| `ts-shared/env-contract/index.test.mts` | Existing | Affected only if implementation evidence requires adding an application-level OpenRouter env declaration | Current AI credentials obey the shared environment contract |

## New tests and scenarios

| Scenario | Setup | Expected outcome |
| -------- | ----- | ---------------- |
| Valid mixed primitive response | Inject deterministic TypeSafe and OpenRouter responses for Noul, Choice, and Score | Exact typed answers in request order, native probabilities/confidence, provider metadata, usage, and raw body are retained |
| Noul confidence invariant | Provider returns a valid Noul probability | Result exposes probability and no synthesized confidence field |
| Exact answer identity | Return missing, duplicate, unknown, reordered, or wrong-primitive answer IDs | Reordering normalizes to request order; every other mismatch rejects the whole call |
| Request and numeric validation | Supply invalid/empty question definitions, invalid Choice criteria, invalid Score criteria, NaN/non-number/out-of-range values, a Choice outside its criteria, a Score outside its level-index range, mismatched Score legends/descriptions/probability keys, or sums outside tolerance | No invalid request dispatches; no invalid response returns any partial result |
| Provider/body failures | Return malformed JSON, unexpected content type/body, ordinary 4xx, 429, 529, and 5xx responses | Malformed/non-retryable failures reject once; retryable failures retry within the bounded policy then surface a redacted error |
| Cancellation | Abort before a request and during retry delay | No further request is issued and the abort propagates |
| Credentialed native contract | With only `OPENROUTER_API_KEY`, call OpenRouter Decisions with all three primitives and pin TypeSafe/no provider fallback | Exactly one live test verifies the concrete Jev model and TypeSafe provider return native decision fields and exact IDs |
| Capacity benchmark | Run repeated fixed-anchor trials with fixed moderation and increasing dynamic sets | Private report clearly says either measured first failure/ceiling or successful through operator cap, never relabels the cap as a ceiling, and contains no credentials/raw content |
| Cost model | Feed observed usage and documented Jev/nano rates and cache-hit assumptions | Deterministic totals expose fixed-set and dynamic-set crossover assumptions |

## Documentation

| Document | Update |
| -------- | ------ |
| `backend/modules/structured-decisions/README.md` | Public API, direct transport contracts, explicit selection, validation/retry invariants, credential injection, and test/benchmark commands |
| `backend/catalogs/README.md` | Register and link the new provider-neutral module |
| `docs/overview/architecture/structured-decisions.md` | Explain the two native decision transports, fail-closed semantics, benchmark method and non-sensitive conclusions, and boundaries with C2/C3; exact operational values remain in the private tier |
| `.env.example`, `docs/development/local-env-vars.md`, `docs/overview/infrastructure/reference-environment-variables-ai-ml.md` | Document `OPENROUTER_API_KEY` setup and that direct TypeSafe credentials are caller-injected/not configured in CI |
| `.github/workflows/VITEST.md` | Regenerate the Vitest project/credential ownership table |

## Verification steps

- `pnpm exec vitest run --project backend-modules backend/modules/structured-decisions`
- `pnpm exec vitest run --project github-actions`
- `pnpm exec vitest run --project ci-tools`
- `pnpm exec vitest run --project ts-shared ts-shared/env-contract/index.test.mts`
- `pnpm run typecheck:backend`
- `pnpm run typecheck:scripts`
- `pnpm exec oxlint --type-aware --deny-warnings backend/modules/structured-decisions`
- `actionlint -ignore 'unexpected key "queue" for "concurrency" section' .github/workflows/tests-backend-credentialed.yml .github/workflows/ci.yml`
- `pnpm run no-mistakes`
- `./ci/lint-links.sh --offline`
- Source `~/voucha.env`, invoke the newly registered `backend-openrouter` project explicitly, and verify the Vitest summary reports one executed test and zero skipped tests.

## Live browser preflight

- Status: `not-required`

## Planning review

- Independent exploration: A read-only codebase explorer traced the module boundary, HTTP dispatcher/retry patterns, workspace registration, Vitest ownership, workflow secret wiring, and deterministic-versus-live test split.
- Independent advisor/reviewer: A separate read-only reviewer audited this Plan against #214, provider contracts, docs-tier policy, Vitest topology, and transient-retry ownership.
- Finding: The first draft needed runtime request validation and precise Score invariants, private handling for operational benchmark values, repeated-anchor ceiling semantics, `backend-modules` exclusion, explicit no-skip live execution, and fifth-project transient ownership. Those corrections are incorporated above.
- Disposition: accepted because the corrected Plan keeps provider data strict, test ownership explicit, operational measurements private, and the implementation bounded to C1.

### PR DESCRIPTION
**Pull request by jonathanong**
## Summary

- Adds a provider-neutral Jev client for Noul, Choice, and Score decisions across direct TypeSafe System One and OpenRouter Decisions.
- Fails closed on request, identity, cardinality, primitive, probability, confidence, provider, and model mismatches while retaining the raw provider response for audit.
- Adds bounded same-model retries, private benchmark tooling, durable architecture documentation, and one isolated credentialed OpenRouter contract test.

## Root cause

The classifier initiative had no reusable structured-decision boundary. The two supported transports expose independently versioned native decision envelopes, and treating either as ordinary generated text would discard the calibrated probabilities and confidence needed by downstream classifiers. Credentialed CI also had no isolated OpenRouter suite able to prove the live native contract without mixing external-provider tests into deterministic backend coverage.

## Architecture

```mermaid
flowchart LR
  Classifier[Future classifier call layer] --> Client[Structured decision client]
  Client -->|direct| TypeSafe[TypeSafe System One]
  Client -->|TypeSafe pinned| OpenRouter[OpenRouter Decisions]
  TypeSafe --> Decoder[Strict decoder]
  OpenRouter --> Decoder
  Decoder --> Result[Typed decisions and raw audit envelope]
```

## Validation

- `pnpm exec vitest run --project backend-modules backend/modules/structured-decisions`
- `pnpm --filter @voucha/backend typecheck`
- `pnpm exec vitest run --project github-actions`
- `pnpm exec vitest run --project ci-tools`
- `pnpm exec vitest run --project backend-openrouter` with the local credential: one test executed, zero skipped
- `pnpm exec oxlint --deny-warnings --type-aware ...` for changed TypeScript files
- `pnpm exec oxfmt --check ...` for all changed files
- `pnpm run no-mistakes`
- `pnpm run syncpack:lint`
- `actionlint` for the changed workflows
- The representative private benchmark completed fixed and dynamic candidate shapes, and its mode-0600 artifact passed schema validation. Exact operational, pricing, usage, cost, and latency observations are intentionally not committed.
Workspace setup: existing non-main worktree; dependencies installed against the committed lockfile.
Agent: codex thread 01a0baf6-2b06-72c0-b5d5-a89f6d2ef556
Device: jonathanong@jonathans-mac-mini.local
Worktree: 1

## Deploy safety

- No existing moderation, tagging, community, or story-clustering path is cut over by this PR.
- `OPENROUTER_API_KEY` is consumed only by the isolated credentialed project and explicit callers.
- OpenRouter requests pin the TypeSafe provider and reject provider or model look-alikes; no cross-model or cross-provider fallback occurs.
- Missing, partial, malformed, or ambiguous responses fail before any typed result is returned.
- The direct TypeSafe path remains fixture-tested until a TypeSafe credential is available for live contract verification.

## Rollout and follow-ups

- #216 will build the multi-candidate call and bounded sharding layer on this client.
- Direct TypeSafe live equivalence and the maximum provider ceiling remain evidence-gated and are not claimed by this change.

## Related issues

Closes #214
Closes #242

<details>
<summary>Shepherd Journal</summary>

- Split primitive answer decoding into a focused module after CI reported SCC complexity 57 > 50; targeted SCC analysis and all 29 deterministic structured-decision tests now pass.
- Refactored the credentialed benchmark into a thin operator entrypoint plus deterministic runner/config/cost units; focused coverage now executes all changed decision, benchmark, retry, validation, and cost paths without provider or filesystem mocks.
- Wrapped benchmark file persistence behind one helper so top-level orchestration remains below the repository sequential-await guard; full ast-grep now passes.
- Accepted Codex review thread PRRT_kwDOUOtx_86kF5_W: the live OpenRouter probe now follows repository credentialed-test convention by skipping when the optional local credential is absent; trusted CI still supplies the configured repository secret and executes the dedicated backend-openrouter project.
- Rebased onto main at 7d328c6a after the base advanced. The only conflict was the transient-infrastructure reference; retained main’s historical reference structure and reapplied the OpenRouter-specific classification boundary. Revalidated 49 deterministic tests, backend typecheck, ast-grep, and the single live OpenRouter probe before force-with-lease.
</details>

### CODE REVIEW

File: backend/modules/structured-decisions/structured-decisions.openrouter.test.mts [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import { describe, expect, it } from 'vitest'
+import { createStructuredDecisionClient } from './structured-decisions.mts'
+
+describe('OpenRouter Decisions', () => {
+  it('preserves native Jev Noul, Choice, and Score fields', async () => {
+    const apiKey = process.env.OPENROUTER_API_KEY
+    if (!apiKey) throw new Error('OPENROUTER_API_KEY is required for backend-openrouter tests.')
```
[the remark this answers is not in the document]
Comment (jonathanong) [reply backend/modules/structured-decisions/structured-decisions.openrouter.test.mts]: Updated the OpenRouter probe to skip when the optional credential is absent while retaining execution in the dedicated credentialed CI project.
